### DISCUSSION TITLE
Query unwanted behaviour in lucid orm and raw sql.

### DISCUSSION THREAD

**Discussion by Vishwas-verma** [adonisjs/core#2874] [Help]
Hey I found something in adonis which is very strange from my end-

The issue that's happening is when I write query using **lucid** i.e
```
    return Driver.query().innerJoin("users", "drivers.user_id", "users.id").innerJoin("bookings", "drivers.id", "bookings.driver_id").innerJoin("reviews", "bookings.id", "reviews.booking_id").groupBy("drivers.id").select("drivers.id", "drivers.user_id", "users.name", "users.email", "users.phone_number").avg({avg_rating: "reviews.rating"}).orderBy("avg_rating", "desc");

```
**It only give me response-**
```
{
    "data": [
        {
            "id": 57,
            "user_id": 170
        },
        {
            "id": 55,
            "user_id": 168
        }
    ]
}
```
 While when I run the sql generated by query as rawQuery i.e
```
   return Database.rawQuery(
      "select `drivers`.`id`,\n" +
      "       `drivers`.`user_id`,\n" +
      "       `users`.`name`,\n" +
      "       `users`.`email`,\n" +
      "       `users`.`phone_number`,\n" +
      "       avg(`reviews`.`rating`) as `avg_rating`\n" +
      "from `drivers`\n" +
      "         inner join `users` on `drivers`.`user_id` = `users`.`id`\n" +
      "         inner join `bookings` on `drivers`.`id` = `bookings`.`driver_id`\n" +
      "         inner join `reviews` on `bookings`.`id` = `reviews`.`booking_id`\n" +
      "group by `drivers`.`id`\n" +
      "order by `avg_rating` desc"
    );
```

Response is-
```
{
    "data": [
        {
            "id": 57,
            "user_id": 170,
            "name": "Dan Kub",
            "email": "Wallace_Marquardt18@gmail.com",
            "phone_number": "705-208-6326 x781",
            "avg_rating": 5
        },
        {
            "id": 55,
            "user_id": 168,
            "name": "Mike Orn",
            "email": "Americo81@hotmail.com",
            "phone_number": "(709) 286-0484 x345",
            "avg_rating": 3.666667
        }
    ]
}
```

How can two query (one derived from other) with same sql produces two different response?
Though Issue I can see is when I write users attributes in **select** statement in adonis query its not showing me those attribute only **parent** model attributes are show i.e drivers .
But it should give some kind of error if its not supported or else produce different sql , sql is generated as required but result is incomplete.
Correct me if I am missing anything in this ?

**Comment by Vishwas-verma** [comment]
> Hey I found something in adonis which is very strange from my end-
> 
> 
> 
> The issue that's happening is when I write query using **lucid** i.e
> 
> ```
> 
>     return Driver.query().innerJoin("users", "drivers.user_id", "users.id").innerJoin("bookings", "drivers.id", "bookings.driver_id").innerJoin("reviews", "bookings.id", "reviews.booking_id").groupBy("drivers.id").select("drivers.id", "drivers.user_id", "users.name", "users.email", "users.phone_number").avg({avg_rating: "reviews.rating"}).orderBy("avg_rating", "desc");
> 
> 
> 
> ```
> 
> **It only give me response-**
> 
> ```
> 
> {
> 
>     "data": [
> 
>         {
> 
>             "id": 57,
> 
>             "user_id": 170
> 
>         },
> 
>         {
> 
>             "id": 55,
> 
>             "user_id": 168
> 
>         }
> 
>     ]
> 
> }
> 
> ```
> 
>  While when I run the sql generated by query as rawQuery i.e
> 
> ```
> 
>    return Database.rawQuery(
> 
>       "select `drivers`.`id`,\n" +
> 
>       "       `drivers`.`user_id`,\n" +
> 
>       "       `users`.`name`,\n" +
> 
>       "       `users`.`email`,\n" +
> 
>       "       `users`.`phone_number`,\n" +
> 
>       "       avg(`reviews`.`rating`) as `avg_rating`\n" +
> 
>       "from `drivers`\n" +
> 
>       "         inner join `users` on `drivers`.`user_id` = `users`.`id`\n" +
> 
>       "         inner join `bookings` on `drivers`.`id` = `bookings`.`driver_id`\n" +
> 
>       "         inner join `reviews` on `bookings`.`id` = `reviews`.`booking_id`\n" +
> 
>       "group by `drivers`.`id`\n" +
> 
>       "order by `avg_rating` desc"
> 
>     );
> 
> ```
> 
> 
> 
> Response is-
> 
> ```
> 
> {
> 
>     "data": [
> 
>         {
> 
>             "id": 57,
> 
>             "user_id": 170,
> 
>             "name": "Dan Kub",
> 
>             "email": "Wallace_Marquardt18@gmail.com",
> 
>             "phone_number": "705-208-6326 x781",
> 
>             "avg_rating": 5
> 
>         },
> 
>         {
> 
>             "id": 55,
> 
>             "user_id": 168,
> 
>             "name": "Mike Orn",
> 
>             "email": "Americo81@hotmail.com",
> 
>             "phone_number": "(709) 286-0484 x345",
> 
>             "avg_rating": 3.666667
> 
>         }
> 
>     ]
> 
> }
> 
> ```
> 
> 
> 
> How can two query (one derived from other) with same sql produces two different response?
> 
> Though Issue I can see is when I write users attributes in **select** statement in adonis query its not showing me those attribute only **parent** model attributes are show i.e drivers .
> 
> But it should give some kind of error if its not supported or else produce different sql , sql is generated as required but result is incomplete.
> 
> Correct me if I am missing anything in this ?

Any update anyone??
@thetutlage

**Reply by feniljariwala82** [reply]
The first thing I noticed is, in the ORM query, you started the query with the "Driver model", which has user_id and id in it. After you are selecting all other fields, which I think won't matter since you have selected the "Driver" class as a parent. 

return **Driver**.query().innerJoin("users", "drivers.user_id", "users.id").innerJoin("bookings", "drivers.id", "bookings.driver_id").innerJoin("reviews", "bookings.id", "reviews.booking_id").groupBy("drivers.id").select("drivers.id", "drivers.user_id", "users.name", "users.email", "users.phone_number").avg({avg_rating: "reviews.rating"}).orderBy("avg_rating", "desc");

But your second raw query looks perfect. Cz your parent table is users, you are starting the joining from the "Users" table. If I am making any mistakes, feel free to advise me!

**Reply by Vishwas-verma** [reply]
Hye I think the second query has driver as the parent model and the most intersting part is second query i.e raw is generated from orm query only. I have converted ORM query to raw and then passed as raw query so they both are basically deriverd from same.

**Reply by feniljariwala82** [reply]
I got your point. Can anybody help us out here?
